### PR TITLE
Pin-based output format selection

### PR DESCRIPTION
**Pull request by ranj063**

### CODE REVIEW

File: sound/soc/sof/ipc4-topology.c [hunk 1]
```diff
@@ -1968,10 +1974,20 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
...
+	 */
+	if (!swidget->prepared) {
+		input_format_index = sof_ipc4_init_input_audio_fmt(sdev, swidget,
+								   &src->base_config,
+								   pipeline_params, available_fmt);
+		if (input_format_index < 0)
+			return ret;
+
+		/* update pipeline memory usage */
+		sof_ipc4_update_resource_usage(sdev, swidget, &src->base_config);
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c]: that looks like a different patch?

I also don't get the comment above, where do we do the ''Otherwise, just update the pipeline params"

File: sound/soc/sof/ipc4-topology.c [hunk 2]
```diff
@@ -1987,7 +2003,7 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 	 * SRC does not perform format conversion, so the output channels and valid bit depth must
 	 * be the same as that of the input.
 	 */
-	in_audio_fmt = &available_fmt->input_pin_fmts[input_format_index].audio_fmt;
+	in_audio_fmt = &src->base_config.audio_fmt;
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c]: not able to review this one, not sure how this is about handling repeated calls when already prepared.

File: sound/soc/sof/ipc4-topology.c [hunk 3]
```diff
@@ -2008,9 +2024,6 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 		return output_format_index;
 	}
 
-	/* update pipeline memory usage */
-	sof_ipc4_update_resource_usage(sdev, swidget, &src->base_config);
-
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c]: I don't understand how you moved this one since the base_config is used in the sof_ipc4_init_output_audio_fmt()??

If you can move it above, let's do it in a separate patch, then deal with the repeated part.

File: sound/soc/sof/sof-audio.c [hunk 4]
```diff
@@ -443,7 +443,7 @@ sof_prepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget
 	if (!widget_ops)
 		return 0;
 
-	if (!swidget || !widget_ops[widget->id].ipc_prepare || swidget->prepared)
+	if (!swidget || !widget_ops[widget->id].ipc_prepare)
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:446]: this is far from self-explanatory, you just add a patch dedicated to filter the multiple calls to prepare and now we remove the test here?

Not clear why we add the if (swidget->prepared) in some places and remove it where it used to be. This feels odd

**Review by plbossart — COMMENTED** [review]
not the simplest PR to finish the day.

File: sound/soc/sof/ipc4-topology.c [hunk 5]
```diff
@@ -1083,13 +1088,43 @@ static int sof_ipc4_init_output_audio_fmt(struct snd_sof_dev *sdev,
 		return 0;
 	}
 
+	/* if module has no output_pin_binding specified in topology, use pin 0 output formats */
```
Comment (bardliao) [inline sound/soc/sof/ipc4-topology.c:1091]: Why pin 0? If there is no output_pin_binding specified, shouldn't we choose the format that matches the reference parms?

Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1091]: @bardliao if there's no output pin binding in the topology, the module likely has identical audio formats for all pins or only has 1 output pin and in both bcases, the audio format definitions in topology will ignore setting the pin_index and by default the pin index will be 0

Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:1091]: But if the topology doesn't specify output pin and each pin can support different formats, we will only use pin 0's format even the widget may not bind to pin 0 with this change.

File: sound/soc/sof/ipc4-topology.c [hunk 6]
```diff
@@ -1065,9 +1065,14 @@ static int sof_ipc4_init_output_audio_fmt(struct snd_sof_dev *sdev,
 					  struct sof_ipc4_base_module_cfg *base_config,
 					  struct sof_ipc4_available_audio_format *available_fmt,
 					  u32 out_ref_rate, u32 out_ref_channels,
-					  u32 out_ref_valid_bits)
```
Comment (bardliao) [inline sound/soc/sof/ipc4-topology.c:1068]: commit message:
"When PCM1 is started" seems to be "When PCM**0** is started,"

**Comment by plbossart** [comment]
@ranj063 I hate to be pushy and I know you're busy, but we're running out of time for reviews before the next merge window. Can you rebase to solve the conflict ?

File: sound/soc/sof/ipc4-topology.c (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1068]: no I meant when PCM1 is started @bardliao. WHen PCM0 is started, things will be fine because we're dealing with PIn 0 already. Its PCM1 that becomes probelmatic

File: sound/soc/sof/ipc4-topology.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1091]: @bardliao ideally when there are multiple outputs, the topology must specify the audio formats for each pin separately even they happen to be identical. But unfortunately thats not what we do today in topology.

**Comment by ranj063** [comment]
> @ranj063 I hate to be pushy and I know you're busy, but we're running out of time for reviews before the next merge window. Can you rebase to solve the conflict ?

@plbossart updated now. Please let me know how it looks.

File: sound/soc/sof/ipc4-topology.c (hunk 3 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c]: it shouldnt be moved. fixed now

File: sound/soc/sof/ipc4-topology.c [hunk 7]
```diff
@@ -1494,6 +1494,53 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 	u32 deep_buffer_dma_ms = 0;
 	int output_fmt_index;
 
+	/* just modify the input params for the next widget if the widget is already prepared */
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:1497]: commit message does not parse for me

In preparation for handling repeated ipc_prepare calls for the same
widget, add the logic to simple update the pipeline params if the
copier is already prepared.

did you mean 'simply update', or maybe 'perform a simple update'

confusion between update as a verb or a noun

File: sound/soc/sof/ipc4-topology.c [hunk 8]
```diff
@@ -1888,12 +1888,14 @@ static int sof_ipc4_prepare_gain_module(struct snd_sof_widget *swidget,
 	u32 out_ref_rate, out_ref_channels, out_ref_valid_bits;
 	int ret;
 
-	ret = sof_ipc4_init_input_audio_fmt(sdev, swidget, &gain->base_config,
-					    pipeline_params, available_fmt);
-	if (ret < 0)
-		return ret;
+	if (!swidget->prepared) {
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:1891]: commit message

````
Add support for handling repeated ipc_prepare calls in the gain, src,
mixer and process type widgets. 
````

You should indicate why we need/want repeated ipc_prepare calls, it's far from obvious for someone like me who keeps forgetting what ipc_prepare is.

````
When the widget is prepared for first
time, the input audio format is initialized in the base_config. 
````
is this good or bad? And what would happen the next time this routine is called?

````
So
replace the use of the index returned from the call to
sof_ipc4_init_input_audio_fmt() to get the input audio to use the
base_config instead.
````

Whaaat?

````
 With repeated ipc_prepare calls, we only need to
update the pipeline params based on the chosen output format.
````
chosen when? the first time, last time?

````
This is only a preparatory step for now as a widget's ipc_prepare callback
be only invoked right now.
````

The grammar does not work for me, did you mean 'callback is only invoked once for now'?

File: sound/soc/sof/sof-audio.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c:446]: The comment still applies, what are we trying to do here?

Also it's not really 'repeated ipc_prepare', what if I understand well is that existing paths will be left alone, it's only the paths that fork that will be impacted.

**Review by plbossart — CHANGES_REQUESTED** [review]
I still find the commit messages way beyond what an average reviewer can understand. Please add more context and explanations, thank you.

File: sound/soc/sof/ipc4-topology.c (hunk 6 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:1068]: > no I meant when PCM1 is started @bardliao. WHen PCM0 is started, things will be fine because we're dealing with PIn 0 already. Its PCM1 that becomes probelmatic

@ranj063 I didn't get "When PCM1 is started, based on the current output format selection logic,
the output format for module-copier.N.1 will be set to S32_LE (based on
the pin 0's output format and the pipeline params will be modified to
reflect this."
Why the output format for module-copier.N.1 will be set based on the pin 0's output format? Pin 0 is not involved, right? The issue looks more like "how do we select out_ref_valid_bits" to me.

**Comment by plbossart** [comment]
@ranj063 can I ask if this PR is still relevant? we are going on almost 2 months and we are going to miss the current kernel cycle at this rate.

**Comment by plbossart** [comment]
closing, no updates since June
